### PR TITLE
REF: implement EA.pad_or_backfill

### LINKED ISSUE #53621: API: EA.ffill/bfill?
ATM EA.fillna handles both a) fillna(scalar), b) ffill/bfill, and c) interpolate (which is mostly broken except for ffill/bfill (which I think should not go through interpolate xref #53581) xref #50950)

At the NDFrame level we have four separate methods: ffill, bfill, fillna, interpolate.  I think it makes sense to just have the same four methods all the way down the stack (Manager, Block, EA).  I'd also be OK with ffill/bfill being combined into e.g. pad_or_backfill for the internal classes.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #53621 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Sits on top of #53989.

The deprecation is a little bit awkward.  If a 3rd party EA has implement fillna but not pad_or_backfill, we warn and call that.  DecimalArray tests that deprecation.

The "method" paths for our existing EA.fillna methods could be ripped out as they are no longer used.  I left them untouched for now.

In a follow-up, can de-duplicate Block.pad_or_backfill with EABlock.pad_or_backfill.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
> /home/runner/work/pandas/pandas/pandas/core/arrays/base.py:docstring of pandas.core.arrays.base.ExtensionArray:119: WARNING: autosummary: stub file not found 'pandas.api.extensions.ExtensionArray.pad_or_backfill'. Check your autosummary_generate setting.

@mroeschke help?  i dont understand what this is asking for

**Comment by jbrockmendel** [comment]
@MarcoGorelli any idea whats going on with the docbuild?  i have a premonition its going to be really obvious as soon as someone points it out to me

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -782,6 +786,85 @@ def interpolate(
...
+    def pad_or_backfill(
+        self,
+        *,
+        method: FillnaOptions,
+        limit: int | None = None,
+        limit_area: Literal["inside", "outside"] | None = None,
+        copy: bool = True,
+    ) -> Self:
+        """
+        pad or backfill values, used by Series/DataFrame ffill and bfill.
```
Comment (mroeschke) [inline pandas/core/arrays/base.py:798]: Yeah I'm not sure if this is related to the docbuild failure, but usually the first letter needs to be capitalized i.e. `Pad or backfill`

Comment (jbrockmendel) [reply pandas/core/arrays/base.py]: hmm that didnt do it

**Comment by MarcoGorelli** [comment]
My guess is that you need to add this new method to doc/source/reference/extensions.rst ?

**Comment by jbrockmendel** [comment]
That solved it, thanks

File: pandas/tests/extension/decimal/array.py [hunk 2]
```diff
@@ -281,6 +281,16 @@ def convert_values(param):
     def value_counts(self, dropna: bool = True):
         return value_counts(self.to_numpy(), dropna=dropna)
 
+    # We override fillna here to simulate a 3rd party EA that has done so. This
```
Comment (mroeschke) [inline pandas/tests/extension/decimal/array.py:284]: So we can remove this in 3.0 when the deprecation is enforced? (if so can we add a note)

Comment (jbrockmendel) [reply pandas/tests/extension/decimal/array.py:284]: Yes.  I'll add that to my next Assorted branch

File: pandas/core/arrays/sparse/array.py [hunk 3]
```diff
@@ -713,6 +713,30 @@ def isna(self):
...
 
+    def pad_or_backfill(
+        self,
+        *,
+        method: FillnaOptions,
+        limit: int | None = None,
+        limit_area: Literal["inside", "outside"] | None = None,
+        copy: bool = True,
+    ) -> Self:
+        msg = "pad_or_backfill with 'method' requires high memory usage."
```
Comment (MarcoGorelli) [inline pandas/core/arrays/sparse/array.py:724]: `'method'` is required here? why not just `"pad_or_backfill for sparse arrays requires high memory usage."`

File: pandas/core/arrays/arrow/array.py [hunk 4]
```diff
@@ -899,6 +899,24 @@ def dropna(self) -> Self:
...
+        self,
+        *,
+        method: FillnaOptions,
+        limit: int | None = None,
+        limit_area: Literal["inside", "outside"] | None = None,
+        copy: bool = True,
+    ) -> Self:
+        if not self._hasna:
+            # TODO(CoW): Not necessary anymore when CoW is the default
+            return self.copy()
```
Comment (MarcoGorelli) [inline pandas/core/arrays/arrow/array.py:912]: is this branch the only reason why `pad_or_backfill` is necessary for arrow array?

**Review by MarcoGorelli — COMMENTED** [review]
looks good, just got minor comments

File: pandas/core/arrays/sparse/array.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:724]: Sure, will update.

File: pandas/core/arrays/arrow/array.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:912]: correct

File: pandas/core/arrays/sparse/array.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py:724]: hmm i think this warning would be unnecessary if we move SparseArray to use the new `get_fill_indexer` implementation.  ill do that as a separate PR

Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py]: updated+green

File: ci/code_checks.sh [hunk 5]
```diff
@@ -82,6 +82,7 @@ if [[ -z "$CHECK" || "$CHECK" == "docstrings" ]]; then
         pandas.api.extensions.ExtensionArray._reduce \
         pandas.api.extensions.ExtensionArray._values_for_factorize \
         pandas.api.extensions.ExtensionArray.interpolate \
+        pandas.api.extensions.ExtensionArray.pad_or_backfill \
```
Comment (MarcoGorelli) [inline ci/code_checks.sh:85]: can we add a little example while we're here?

**Review by MarcoGorelli — COMMENTED** [review]
might need another merge after recent CI changes

I'd suggest adding a little example here too, so as to not add even more to the list of exclusions (which we're hoping to remove next week)

**Review by MarcoGorelli — APPROVED** [review]
Nice one! Ok to merge imo

**Review by mroeschke — APPROVED** [review]

**Pull request merged**